core/sdk-core/src/main/java/software/amazon/awssdk/core/waiters/WaiterAcceptor.java

```diff
@@ -0,0 +1,157 @@
+     */
+    static <T> WaiterAcceptor<T> successOnResponseAcceptor(Predicate<T> responsePredicate) {
+        return new WaiterAcceptor<T>() {
+            @Override
+            public WaiterState waiterState() {
+                return WaiterState.SUCCESS;
+            }
+
+            @Override
+            public boolean matches(T response) {
+                return responsePredicate.test(response);
+            }
+        };
+    }
+
+    /**
+     * Creates an error waiter acceptor which determines if the exception should transition the waiter to success state
+     *
+     * @param errorPredicate the {@link Throwable} predicate
+     * @param <T> the response type
+     * @return a {@link WaiterAcceptor}
+     */
+    static <T> WaiterAcceptor<T> successOnExceptionAcceptor(Predicate<Throwable> errorPredicate) {
+        return new WaiterAcceptor<T>() {
+            @Override
+            public WaiterState waiterState() {
+                return WaiterState.SUCCESS;
+            }
+
+            @Override
+            public boolean matches(Throwable t) {
+                return errorPredicate.test(t);
+            }
+        };
+    }
+
+    /**
+     * Creates an error waiter acceptor which determines if the exception should transition the waiter to failure state
+     *
+     * @param errorPredicate the {@link Throwable} predicate
+     * @param <T> the response type
+     * @return a {@link WaiterAcceptor}
+     */
+    static <T> WaiterAcceptor<T> errorOnExceptionAcceptor(Predicate<Throwable> errorPredicate) {
+        return new WaiterAcceptor<T>() {
+            @Override
+            public WaiterState waiterState() {
+                return WaiterState.FAILURE;
+            }
+
+            @Override
+            public boolean matches(Throwable t) {
+                return errorPredicate.test(t);
+            }
+        };
+    }
+
+    /**
+     * Creates a success waiter acceptor which determines if the exception should transition the waiter to success state
+     *
+     * @param responsePredicate the predicate of the response
+     * @param <T> the response type
+     * @return a {@link WaiterAcceptor}
+     */
+    static <T> WaiterAcceptor<T> errorOnResponseAcceptor(Predicate<T> responsePredicate) {
+        return new WaiterAcceptor<T>() {
+            @Override
+            public WaiterState waiterState() {
+                return WaiterState.FAILURE;
+            }
+
+            @Override
+            public boolean matches(T response) {
+                return responsePredicate.test(response);
+            }
+        };
+    }
+
+    /**
+     * Creates a retry on exception waiter acceptor which determines if the exception should transition the waiter to retry state
+     *
+     * @param errorPredicate the {@link Throwable} predicate
+     * @param <T> the response type
+     * @return a {@link WaiterAcceptor}
+     */
+    static <T> WaiterAcceptor<T> retryOnExceptionAcceptor(Predicate<Throwable> errorPredicate) {
```

Comment:
What about `retryOnResponseAcceptor`?

Reply:
That's fair, but it feels like a weird omission, even if people wouldn't use it very much. A `retryOnResponseAcceptor` would still be useful to allow someone to short-circuit a more broad acceptor later in the list:

E.g.
```java
waiterBuilder.addAcceptor(retryOnResponseAcceptor(r -> r.errorCode() == FieldValues.RETRYABLE_ERROR)
             .addAcceptor(failOnResponseAcceptor(r -> r.errorCode() != null))
```

I agree that it's not necessary if all we're using this for is the generated waiters, but we're writing it more generic for a reason.

I'd almost rather have the default behavior of the core Waiters to *not* retry by default and just have all of the generated waiters use `addAcceptor(retryOnResponseAcceptor(r -> true))` since that seems more of an intuitive default.

Reply:
Yeah, makes sense. I'll add it.